src/mpl/src/gpu/mpl_gpu_cuda.c

```diff
@@ -327,15 +327,17 @@ int MPL_gpu_free_hook_register(void (*free_hook) (void *dptr))
 CUresult CUDAAPI cuMemFree(CUdeviceptr dptr)
 {
     CUresult result;
-    gpu_free_hooks_cb((void *) dptr);
+    if ((void *) dptr != NULL)
+        gpu_free_hooks_cb((void *) dptr);
     result = sys_cuMemFree(dptr);
```

Comment:
Shouldn't this line inside the condition as well?

Reply:
sys_cuMemFree should deal with all input no matter whether it is NULL or not

Reply:
The question is what does `sys_cuMemFree(NULL)` do?

I understand that the function internally may check and skip `NULL`, but that is internal and *non-obvious* behavior. The upper layer is *assuming* this *non-obvious* behavior. This kind of code based *assumption* is bad. It is the very reason that we had this bug in the first place -- that we assumed `gpu_free_hooks_cb` handles `NULL`.

So if we learned our lesson, than we should replace these kind of *assumptions* with either explicit condition or explicit comments.

Reply:
How do you know `sys_cuMemFree` will internally check and skip `NULL`? It is close source.
`sys_cuMemFree` is cuda free function in cuda library, NULL to this function is valid input. For this hook, the only thing we add here is `gpu_free_hooks_cb`; but `gpu_free_hooks_cb` cannot deal with NULL, that's why the bug happens.
However, now matter whether the input pointer is NULL or not, we should always pass it to cuda free function `sys_cuMemFree`

Reply:
Why we call and what does `sys_cuMemFree(NULL)` supposed to do? I am asking for semantics here. Closed-source has nothing to do with it. We have to know what a function does before using that function. If you have documentation, please quote reference.

Why I make it an issue? Because "memfree" a NULL address doesn't make sense to me and a code that doesn't make sense is bad code.

Reply:
> the only thing we add here is `gpu_free_hooks_cb`; but `gpu_free_hooks_cb` cannot deal with NULL, that's why the bug happens.

I understand this. If the original code is like following:
```
if (dptr) {
    result = sys_cuMemFree(dptr);
}
```
You would never make the mistake of 
```
gpu_free_hooks_cb(dptr);
if (dptr) {
    result = sys_cuMemFree(dptr);
}
```
Right?

The original code was bad code, that is why enticed you to create the bug in the first place. Now we know the code was bad, we should fix it -- unless I misunderstood the code, then we need add comments to explain and prevent the future misunderstanding.

Reply:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1ga042655cbbf3408f01061652a075e094
says if input is 0, no operation is done. `NULL` is valid input for `cudaFree`, we should always pass it to `cudaFree` in cuda library in our hook. If in future Nvidia decides to do something with NULL pointer, our code can work well at that time

Reply:
so what is your suggestion here? We should do this?
```
if (dptr) {
    gpu_free_hooks_cb(dptr);
    result = sys_cuMemFree(dptr);
}
```

Reply:
> says if input is 0, no operation is done

Then why we call the function if we know the pointer is NULL?

> If in future Nvidia decides to do something with NULL pointer, our code can work well at that time

That is non-sesical. A NULL pointer doesn't belong to CUDA, it is not created by CUDA, and Cuda can't decide suddenly to do something with NULL. Right?

Reply:
what if cuda want to warn users
`Warnning: free with NULL pointer` in future?

Reply:
> 
> 
> so what is your suggestion here? We should do this?
> 
> ```
> if (dptr) {
>     gpu_free_hooks_cb(dptr);
>     result = sys_cuMemFree(dptr);
> }
> ```

Absolutely.

Reply:
> 
> 
> what if cuda want to warn users
> `Warnning: free with NULL pointer` in future?

You are saying we want write code that do nothing with the only purpose is to get a warning?

Reply:
or probably even cuda decided if user calls `cudaFree(NULL)`, that becomes invalid operations? Everything could happen in future, right? What I say is we should not decide whether the ptr is passed to cudaFree function or not, not matter what pointer user gives, we should always give it to cudaFree in cuda library. The only thing we need to do is `gpu_free_hooks_cb`.

Reply:
There is this practice called "defensive programming". For example, `free` should only free a pointer that the same library originally allocated. But user may call free on a NULL pointer sometime, which *always* is a user mistake. But mistakes happens, and it is trivial to add a "if"-guard. This `if`-guard is not necessary for the correct `free` implementation, it is there as a double safety measure, to provide more resilience toward mistakes. By nomean, having this `if`-guard is meant to encourage the user to write erroneous code to call free on NULL.

A user who claims, that you should check whether address is NULL because `free` is okay to be called on NULL is a non-sensical programmer.

Reply:
> if user calls `cudaFree(NULL)`, that becomes invalid operations? 

Isn't that an argument that we should *not* call `cudaFree(NULL)`? 

> What I say is we should not decide whether the ptr should be passed to cudaFree function or not, not matter what pointer user gives, we should always give it to cudaFree in cuda library.

Who are the user? Are you freeing a pointer that is not allocated by you? ... that is really bad design.

Reply:
@KaimingOuyang I asked you for semantics but you never replied with semantics. The sematics of cudafree is to free an address that is *allocated* by cuda, right? A NULL pointer is not an address allocated by cuda, right? If you follow the semantics, I don't understand why there is argument. If you are not following or even thinking about semantics, then you should be.

Reply:
ok, let's assume right now I am writing cuda library. My `cudaFree` will give a warning if the user passes NULL pointer, because I don't want to let users think pass a NULL ptr is a good practice. Then, if you are writing a hook like you suggested, if input ptr is NULL, you basically change the behavior of `cudaFree` in cuda library. Do you think this is valid?

Reply:
Let's move offline

Reply:
@KaimingOuyang If you want minimal changes, why don't you move the checking of `NULL` inside `gpu_free_hooks_cb`?

Reply:
since we need to consider `sys_cuMemFree`, it is better to add if-structure in hook function not callback function.